presto-hive/src/main/java/io/prestosql/plugin/hive/util/SerDeUtils.java

```diff
@@ -312,11 +320,28 @@ private static long formatDateAsLong(Object object, DateObjectInspector inspecto
         return inspector.getPrimitiveJavaObject(object).toEpochDay();
     }
 
-    private static long formatTimestampAsLong(Object object, TimestampObjectInspector inspector)
+    private static DecodedTimestamp formatTimestamp(TimestampType type, Object object, TimestampObjectInspector inspector)
     {
+        long epochSecond;
+        int nanoOfSecond;
+
         if (object instanceof TimestampWritable) {
-            return ((TimestampWritable) object).getTimestamp().getTime() * MICROSECONDS_PER_MILLISECOND;
+            TimestampWritable timestamp = (TimestampWritable) object;
+            epochSecond = timestamp.getSeconds();
+            nanoOfSecond = timestamp.getNanos();
         }
-        return inspector.getPrimitiveJavaObject(object).toEpochMilli() * MICROSECONDS_PER_MILLISECOND;
+        else {
+            Timestamp timestamp = inspector.getPrimitiveJavaObject(object);
+            epochSecond = timestamp.toEpochSecond();
+            nanoOfSecond = timestamp.getNanos();
+        }
+
+        nanoOfSecond = (int) round(nanoOfSecond, 9 - type.getPrecision());
+        if (nanoOfSecond == NANOSECONDS_PER_SECOND) { // round nanos up to seconds
+            epochSecond += 1;
+            nanoOfSecond = 0;
+        }
```

Comment:
This is probably a change in behavior, but probably a correct one. 
The old code
```
((TimestampWritable) object).getTimestamp().getTime() * MICROSECONDS_PER_MILLISECOND
```
```
inspector.getPrimitiveJavaObject(object).toEpochMilli() * MICROSECONDS_PER_MILLISECOND;
```

was probably truncating.

Am i correct?
Does it apply to timestamp outside of structural types as well?
Do we have a test for this?

Reply:
Yes, the original version did truncate.

>  Does it apply to timestamp outside of structural types as well?

No; the only public API in `SerdeUtils` is `getBlockObject`<sup>1</sup>, which is only used in `GenericHiveRecordCursor.parseObjectColumn`, and only when `type.getJavaType() == Block.class`.

>  Do we have a test for this?

We did not previously have a  test for this (the only tests I can find I in five minutes for arrays in Hive is in `TestHiveIntegrationSmokeTest`). The tests added in this PR do cover it, though (because they test all available precisions).

We should probably have a class like `TestHiveTypes`. Should I open an issue for that?

---

<sup>1</sup> `serializeObject` is also public, but it looks like it might be missing `@VisibleForTesting`. It is only used in `SerDeUtils` and tests, and its overload is `@VisibleForTesting`.